File: Day71/Path_Sum_3.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    
    int pathSum(TreeNode* root, int targetSum) 
    {
        stack<TreeNode*> nodes;
        nodes.push(root);
        int count = 0;

        while(!nodes.empty())
        {
            TreeNode* top = nodes.top();
            nodes.pop();

            if(top != NULL)
            {
                nodes.push(top->left);
                nodes.push(top->right);
            }
            getPath(top, 0, targetSum, count);
        }

        return count;
    }

    void getPath(TreeNode* curRoot, long currSum, int targetSum, int &count)
    {
        if(curRoot == NULL)
        {
            return;
        }

        currSum += curRoot->val;
        if(currSum == targetSum)
        {
            count++;
        }

        getPath(curRoot->left, currSum, targetSum, count);
        getPath(curRoot->right, currSum, targetSum, count);
    }
};

File: Day71/Path_Sum3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> int:

        self.count = 0
        #using a stack to insert each and every root node
        stack = []
        stack.append(root)

        def getPath(root, currSum):
            if root == None:
                return

            currSum += root.val

            if currSum == targetSum:
                self.count += 1

            getPath(root.left, currSum)
            getPath(root.right, currSum)

        while stack:
            root = stack.pop()
            if root:
                #insert the child nodes of root into stack
                stack.append(root.left)
                stack.append(root.right)

            getPath(root, 0)

        return self.count





        
